ci-operator/step-registry/cucushift/upgrade/rollback/cucushift-upgrade-rollback-commands.sh

```diff
@@ -0,0 +1,140 @@
+#!/bin/bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+trap 'FRC=$?; createUpgradeJunit; debug' EXIT TERM
+
+export HOME="${HOME:-/tmp/home}"
+export XDG_RUNTIME_DIR="${HOME}/run"
+export REGISTRY_AUTH_PREFERENCE=podman # TODO: remove later, used for migrating oc from docker to podman
+mkdir -p "${XDG_RUNTIME_DIR}"
+# After cluster is set up, ci-operator make KUBECONFIG pointing to the installed cluster,
+# to make "oc registry login" interact with the build farm, set KUBECONFIG to empty,
+# so that the credentials of the build farm registry can be saved in docker client config file.
+# A direct connection is required while communicating with build-farm, instead of through proxy
+KUBECONFIG="" oc --loglevel=8 registry login
+
+# Print cv, failed node, co, mcp information for debug purpose
+function debug() {
+    if (( FRC != 0 )); then
+        echo -e "oc get clusterversion/version -oyaml\n$(oc get clusterversion/version -oyaml)"
+        echo -e "oc get machineconfig\n$(oc get machineconfig)"
+        echo -e "Describing abnormal nodes...\n"
+        oc get node --no-headers | awk '$2 != "Ready" {print $1}' | while read node; do echo -e "\n#####oc describe node ${node}#####\n$(oc describe node ${node})"; done
+        echo -e "Describing abnormal operators...\n"
+        oc get co --no-headers | awk '$3 != "True" || $4 != "False" || $5 != "False" {print $1}' | while read co; do echo -e "\n#####oc describe co ${co}#####\n$(oc describe co ${co})"; done
+        echo -e "Describing abnormal mcp...\n"
+        oc get mcp --no-headers | awk '$3 != "True" || $4 != "False" || $5 != "False" {print $1}' | while read mcp; do echo -e "\n#####oc describe mcp ${mcp}#####\n$(oc describe mcp ${mcp})"; done
+    fi
+}
+
+# Generate the Junit for upgrade
+function createUpgradeJunit() {
+    echo "Generating the Junit for upgrade"
+    if (( FRC == 0 )); then
+      cat >"${ARTIFACT_DIR}/junit_upgrade.xml" <<EOF
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuite name="cluster upgrade" tests="1" failures="0">
+  <testcase classname="cluster upgrade" name="upgrade should succeed"/>
+</testsuite>
+EOF
+    else
+      cat >"${ARTIFACT_DIR}/junit_upgrade.xml" <<EOF
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuite name="cluster upgrade" tests="1" failures="1">
+  <testcase classname="cluster upgrade" name="upgrade should succeed">
+    <failure message="">openshift cluster upgrade failed</failure>
+  </testcase>
+</testsuite>
+EOF
+    fi
+}
+
+function run_command() {
+    local CMD="$1"
+    echo "Running command: ${CMD}"
+    eval "${CMD}"
+}
+
+# Rollback the cluster to target release
+function rollback() {
+    res=$(run_command "oc adm upgrade rollback")
+    out="Requested rollback from ${TARGET_VERSION} to ${SOURCE_VERSION}"
+    if [[ ${res} == *"${out}"* ]]; then
+        echo "Rolling back cluster gets started..."
+    else
+        echo "Rolling back cluster doesn't start..."
```

Comment:
perhaps we should fail (return 1) if we're getting unexpected output, as well as capture stderr (2>&1) into res and print the error vs the expected? wdyt?
```suggestion
    res=$(oc adm upgrade rollback 2>&1 || true)
    out="Requested rollback from ${TARGET_VERSION} to ${SOURCE_VERSION}"
    if [[ ${res} == *"${out}"* ]]; then
        echo "Rolling back cluster from ${TARGET_VERSION} to ${SOURCE_VERSION} started..."
    else
        echo "Rolling back cluster returned unexpected:\n${res}\nexpecting: ${out}"
        return 1
```

Reply:
Updated.